yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+        setLastFailure(results.failureCause.get, results.timestamp)
+        // and the failure counter
+        refreshFailedCount.incrementAndGet()
+      }
+    }
+    results
+  }
+
+  /**
+   * List applications. This currently finds completed applications only.
+   * 
+   * If the timeline is  not enabled, returns an empty list
+   * @return List of all known applications.
+   */
+  override def getListing(): Seq[ApplicationHistoryInfo] = {
+    // get the current list
+    val listing = getApplications().applications
+    // and queue another refresh
+    triggerRefresh()
+    listing
+  }
+
+  /**
+   * Trigger a refresh
+   */
+  def triggerRefresh(): Unit = {
+    refresher.refresh(now())
+  }
+
+  /**
+   * Return the current time
+   * @return
+   */
+  def now(): Long = {
+    System.currentTimeMillis()
+  }
+
+  /**
+   * Get the last refresh attempt (Which may or may not be successful)
+   * @return the last refresh time
+   */
+  def getLastRefreshAttemptTime(): Long = {
+    refresher.lastRefreshAttemptTime
+  }
+  
+  /**
+   * Look up the timeline entity
+   * @param appId application ID
+   * @return the entity associated with the given application
+   * @throws FileNotFoundException if no entry was found
+   */
+  def getTimelineEntity(appId: String): TimelineEntity = {
+    logDebug(s"GetTimelineEntity $appId")
+    maybeCheckHealth()
+    getTimelineQueryClient().getEntity(YarnHistoryService.SPARK_EVENT_ENTITY_TYPE, appId)
+  }
+
+
+  /**
+   * Returns the Spark UI for a specific application.
+   * <p>
+   * If the timeline is  not enabled, returns `None`
+   * @param appId The application ID.
+   * @param attemptId The application attempt ID (or `None` if there is no attempt ID).
+   * @return The application's UI, or `None` if application is not found.
+   */
+  override def getAppUI(appId: String, attemptId: Option[String]): Option[SparkUI] = {
+    getAppUI(appId)
+  }
+
+  /**
+   * Build the application UI for an application
+   * <p>
+   * If the timeline is  not enabled, returns `None`
+   * @param appId The application ID.
+   * @return The application's UI, or `None` if application is not found.
+   */
+  def getAppUI(appId: String): Option[SparkUI] = {
+    logDebug(s"Request UI with appId $appId")
+    if (!enabled) {
+      // Timeline is disabled: return nothing
+      return None
+    }
+    maybeCheckHealth()
+    try {
+      val entity = getTimelineEntity(appId)
+
+      if (log.isDebugEnabled) {
+        logDebug(describeEntity(entity))
+      }
+      val bus = new SparkListenerBus() {}
+      val appListener = new ApplicationEventListener()
+      bus.addListener(appListener)
+
+      val ui = {
+        val conf = this.sparkConf.clone()
+        val appSecManager = new SecurityManager(conf)
+        SparkUI.createHistoryUI(conf, bus, appSecManager, appId,
+                                 HistoryServer.UI_PATH_PREFIX + s"/${appId }", entity.getStartTime)
+      }
+      val events = entity.getEvents
```

Comment:
1. There's no partial retrieval in the ATS APIs; the assumption in the timeline server is that you don't call this often so there's less cost on the server. 
2. If you ever look at where CPU load goes on the server when making these (or any of the timeline server calls), marshalling is a main expense. Maybe we should push for a thrift or PB format option in future.
3. There's an implicit assumption in the history server that the cache fetch lookup is fast and doesn't fail; something async in the UI with some meta-refresh could address that in the HTML view, but not JSON. 
4. Note that [SPARK-7889] unintentionally reduces the load placed on the timeline server by in-progress applications.